Reduce unnecessary wrapping with nested `new()`: `InplaceBox::new(InplaceBox::new())`

drmingdrmer opened the pull request:
- Fix: #4

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/drmingdrmer/inplace-box/5)
<!-- Reviewable:end -->

drmingdrmer committed:
chore: move new() to separate file

drmingdrmer reviewed (commented):
@schreter 
This approached does not seem to work for the following two restriction :(

I've tried other ways such as put specializing on `InplaceBox`, but still can not work. Because `InplaceBox` does not implement `Unsize`.

- See: #6

> *[Reviewable](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-:-OWAdTYy9HXazTIhEtTw:b-31vb4g)* status: 0 of 2 files reviewed, 2 unresolved discussions

___
*[`src/impl_new.rs` line 54 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAdASv3sNA9p_SMfuW:-OWAdASv3sNA9p_SMfuX:b2itiz7) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L54)):*
> ```XML
>         'a,
>         T: ?Sized + 'a,
>         U: Sized + Unsize<T> + IsInPlaceBox<T, SIZE> + 'a,
> ```

The second issue is that `min_specialization` does not apply to user defined trait, `IsInPlaceBox<...>`:

```
error: cannot specialize on trait `IsInPlaceBox`
  --> src/impl_new.rs:54:32
   |
54 |         U: Sized + Unsize<T> + IsInPlaceBox<T, SIZE> + 'a,
```
___
*[`src/impl_new.rs` line 75 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAd-4DE4c1S6F0gren:-OWAd-4DE4c1S6F0greo:bbb0ej4) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L75)):*
> ```XML
>         fn assert_unsized<U: ?Sized, T: Unsize<U>>(t: &T) {}
>         let b1: InplaceBox<dyn Trait, 8> = InplaceBox::new(42u64);
>         assert_unsized::<dyn Trait, _>(&b1);
> ```

This approach does not work because 1): InplaceBox does not implement `Unsize` thus it can not be specialized.

```
error[E0277]: the trait bound `InplaceBox<dyn Trait, 8>: Trait` is not satisfied
  --> src/impl_new.rs:75:37
   |
75 |         assert_unsized::<dyn Trait, _>(&b1);
   |                                     ^ the trait `Trait` is not implemented for `InplaceBox<dyn Trait, 8>`
   |
   = help: the trait `Trait` is implemented for `u64`
```



_Code quote:_
```XML
    trait Trait {}
    impl Trait for u64 {}

    #[test]
    fn test_inplace_box_unsized() {
        fn assert_unsized<U: ?Sized, T: Unsize<U>>(t: &T) {}
        let b1: InplaceBox<dyn Trait, 8> = InplaceBox::new(42u64);
        assert_unsized::<dyn Trait, _>(&b1);
```


<!-- Sent from Reviewable.io -->

schreter reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/drmingdrmer/inplace-box/5)* status: all files reviewed, 2 unresolved discussions (waiting on @drmingdrmer)

___
*[`src/impl_new.rs` line 54 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAdASv3sNA9p_SMfuW:-OWDECeFEEa-B4syPNvE:bnc82zs) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L54)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

The second issue is that `min_specialization` does not apply to user defined trait, `IsInPlaceBox<...>`:

```
error: cannot specialize on trait `IsInPlaceBox`
  --> src/impl_new.rs:54:32
   |
54 |         U: Sized + Unsize<T> + IsInPlaceBox<T, SIZE> + 'a,
```
</blockquote></details>

Aaah... You are right, I was using `specialization`, not `min_specialization` :-(. Sorry for the confusion. Not sure if we want to have `specialization` feature, probably not.

However, what *should* be possible is to cast to `Any` and then `downcast_ref()` to `InplaceBox<T, SIZE>` and/or compare `TypeId` of the two. If it is `InplaceBox`, then just copy 1:1, else do the regular thing. The problem is the static assertion - it will fire every time. If the `TypeId` comparison works, then it could be added to the assertions. Now it's waaaay too late and not sure I can look into it tomorrow, so just as a pointer.


<!-- Sent from Reviewable.io -->

drmingdrmer committed:
feat: add specialized new() for nested `InplaceBox::new`

Implement trait-based specialization to optimize InplaceBox creation:
- Move existing InplaceBox values without conversion
- Regular values use standard in-place construction

Such implementation requires the specification feature flag.

The type parameter `U` is removed from `ConvertIntoInplaceBox`.

drmingdrmer reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/drmingdrmer/inplace-box/5)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @schreter)

___
*[`src/impl_new.rs` line 54 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAdASv3sNA9p_SMfuW:-OWE3otDD_qTy_iqKBtN:bao0e19) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L54)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

Aaah... You are right, I was using `specialization`, not `min_specialization` :-(. Sorry for the confusion. Not sure if we want to have `specialization` feature, probably not.

However, what *should* be possible is to cast to `Any` and then `downcast_ref()` to `InplaceBox<T, SIZE>` and/or compare `TypeId` of the two. If it is `InplaceBox`, then just copy 1:1, else do the regular thing. The problem is the static assertion - it will fire every time. If the `TypeId` comparison works, then it could be added to the assertions. Now it's waaaay too late and not sure I can look into it tomorrow, so just as a pointer.
</blockquote></details>

`specialization` looks good. It's said the main issue about it is issues with lifetime.
___
*[`src/impl_new.rs` line 36 at r2](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWE3bOnCxP9MlUV3ZaO:-OWE3bOnCxP9MlUV3ZaP:b-vh7pk6) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/63a32a684598c49a7e7dd16323f4b149863be7ef/src/impl_new.rs#L36)):*
> ```XML
> /// Provides specialized behavior: regular values are placed in-place,
> /// existing `InplaceBox` values are moved without conversion.
> trait ConvertIntoInplaceBox<'a, T, const SIZE: usize>
> ```

I removed type `U`  from `ConvertIntoInplaceBox`. It seems unnecessary. Or am I missing something?


<!-- Sent from Reviewable.io -->

schreter reviewed (commented):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/drmingdrmer/inplace-box/5)* status: all files reviewed, 1 unresolved discussion (waiting on @drmingdrmer)

___
*[`src/impl_new.rs` line 54 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAdASv3sNA9p_SMfuW:-OWEmph12m9cWhHkqFGd:b-r4a172) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L54)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

`specialization` looks good. It's said the main issue about it is issues with lifetime.
</blockquote></details>

OK, then let's go with `specialization`. I'm just normally cautious with features which are not 100% complete :-). But I think `specialization` will definitely come, since it's plain needed.
___
*[`src/impl_new.rs` line 75 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAd-4DE4c1S6F0gren:-OWEn2ljDl4Gf9jfHCvL:b-p4j6iv) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L75)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

This approach does not work because 1): InplaceBox does not implement `Unsize` thus it can not be specialized.

```
error[E0277]: the trait bound `InplaceBox<dyn Trait, 8>: Trait` is not satisfied
  --> src/impl_new.rs:75:37
   |
75 |         assert_unsized::<dyn Trait, _>(&b1);
   |                                     ^ the trait `Trait` is not implemented for `InplaceBox<dyn Trait, 8>`
   |
   = help: the trait `Trait` is implemented for `u64`
```
</blockquote></details>

This should be gone now, right?
___
*[`src/impl_new.rs` line 36 at r2](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWE3bOnCxP9MlUV3ZaO:-OWEmlCe7MI5ijnAzyIl:bt6y286) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/63a32a684598c49a7e7dd16323f4b149863be7ef/src/impl_new.rs#L36)):*
<details><summary><i>Previously, drmingdrmer (张炎泼) wrote…</i></summary><blockquote>

I removed type `U`  from `ConvertIntoInplaceBox`. It seems unnecessary. Or am I missing something?
</blockquote></details>

Yes, you are right, this doesn't seem needed.
___
*[`src/lib.rs` line 61 at r2](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWEn8jeD3nhfVB1tQjo:-OWEn8jeD3nhfVB1tQjp:b5x9q6k) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/f752593f6bdd5946434f88f545fa9448aad081dc/src/lib.rs#L61)):*
> ```XML
>     ///
>     /// Type match, size and alignment are checked statically by the compiler.
>     pub fn new<'a, U: Sized + Unsize<T> + 'a>(value: U) -> Self {
> ```

We still need `new()`, but instead of in-place implementation, to delegate to a trait. How comes the tests below with `new()` work?


<!-- Sent from Reviewable.io -->

drmingdrmer reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/drmingdrmer/inplace-box/5)* status: all files reviewed, 1 unresolved discussion (waiting on @schreter)

___
*[`src/impl_new.rs` line 75 at r1](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWAd-4DE4c1S6F0gren:-OWEr-35AUrB2kLqtwh_:bfpaqhu) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/d345bb0a805e6e358e95eafd8e4c521343e469d8/src/impl_new.rs#L75)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

This should be gone now, right?
</blockquote></details>

Yes. I've removed this test. You should not see it anymore in the last commit.
___
*[`src/lib.rs` line 61 at r2](https://reviewable.io/reviews/drmingdrmer/inplace-box/5#-OWEn8jeD3nhfVB1tQjo:-OWEqqwC2HIYiN3KbJqt:b-a0r6s0) ([raw file](https://github.com/drmingdrmer/inplace-box/blob/f752593f6bdd5946434f88f545fa9448aad081dc/src/lib.rs#L61)):*
<details><summary><i>Previously, schreter wrote…</i></summary><blockquote>

We still need `new()`, but instead of in-place implementation, to delegate to a trait. How comes the tests below with `new()` work?
</blockquote></details>

I moved new() to `src/impl_new.rs` :thinking: 


<!-- Sent from Reviewable.io -->

drmingdrmer commented:
The latest is published to 0.2.1:
https://crates.io/crates/inplace-box/0.2.1
